### ISSUE TITLE
Allow git lfs migrate import to take a .gitattributes file as an argument

### ISSUE THREAD

**Issue by bk2204** [git-lfs/git-lfs#3442]
In some cases, a user will want to filter one or more branches when they already have a valid .gitattributes file they want to use. Currently, they must turn Git LFS off, rebase the .gitattributes file into the root of the tree, then migrate the commits. It would be a better user experience if the user could specify a .gitattributes file when running git lfs migrate import, so that they can use the existing file.

**Comment by nhatkhai** [comment]
For clarify, I don't have to turn off LFS if want to go from non-LFS to LFS branch. I first do a index-filter or tree-filter and add .gitattrubites file to every commits. Then run git lfs imgrate import --fixup, or --inculde=.... for the older lfs versions. So with --fixup, it kind nicer, but if I can get rich of the index-filer step it would be great.

The is a warning. I noted that for version 4.x lfs, I have to use tree-filter, and lfs migrate inorder to have 100% all of the file that I wanted converted into LFS. If I left out one of them, I left with ~80% time are converted. So, I hope you have some test to me sure this not happen.

**Comment by Rabadash8820** [comment]
Yes, this would be exactly what I need! I have been trying to clean up some old Unity repos to use LFS. These repos have long histories with multiple changes to binary assets, and the `.gitattributes` file wasn't even added until the latest commit because I didn't know about LFS when I started those projects. So I can't use `import --fixup` as is because there is no `.gitattributes` in the history, and I can't just add the `.gitattributes` now by rewriting history, because then I get a bunch of `file should be pointer but wasn't` errors while rebasing.

**Comment by nhatkhai** [comment]
```
git filter-branch --prune-empty -f --index-filter '
  cp -f my_ideal_gitattributes .gitattributes
  git add .gitattributes
' -- HEAD

git lfs migrate import --fixup HEAD
```

**Comment by phobos2077** [comment]
Any news about this? Seems like a must have feature missing. I need to migrate an existing repository to LFS w/o rewriting any history and I have a .gitattributes already configured with a lot of extensions. Do I need to run git lfs migrate a lot of times for each file extension, is there no other way?

**Comment by nhatkhai** [comment]
@phobos2077 - When you run migrate command - it is a rewrite history action (It is the only way I know how git works).

**Comment by bk2204** [comment]
We haven't implemented this yet, but we're happy to accept a patch if you'd like to send one.

If you're not interested in rewriting history, you can add the `.gitattributes` file to the repository and then run `git add --renormalize .` before committing, which should rewrite all the necessary files into LFS objects.  Note that this requires a relatively recent Git.

**Comment by phobos2077** [comment]
@nhatkhai migrate command has a `--no-rewrite` option that I wanted to use which doesn't involve changing history. Rewriting history is not an option because the repository is distributed with many developers working on it. Having to organize everybody to freeze their work, checkout every branch locally, rewrite the entire history of every branch and then somehow re-clone the repository for everybody... That would be insane :)

@bk2204 thanks. I didn't knew about this option. However I've already tried doing `rm --cached **/*.[extension]` which seems to work exactly like I wanted (except I had to type in every extension one by one of course). I've just added everything and committed after running this and it seems to achieve the goal of having repository in a correct state where all files that should be in LFS are there. Now the only problem remaining is to merge this commit from master to every feature branch and resolve any conflicts that will certainly arise because I've just changed every binary file in the project.

Is there any possible issues that I'm not seeing with this approach?

**Comment by bk2204** [comment]
If you use `git rm --cached`, you'll of course need to re-add the objects in order to have them committed.  When you do that, Git will filter them through Git LFS, and you'll get the same result.  `git add --renormalize` essentially runs the processes through smudge and clean states again to get them filtered, which is the same thing you're doing by re-adding them.

So there's no downsides to your approach other than it being less convenient.  `git lfs migrate import --no-rewrite` precedes the existence of `git add --renormalize`, so you can also use it, although as you've noted it is also not as convenient as it could be.

**Comment by nhatkhai** [comment]
@phobos2077, I do not know how --renormalize work. If what you want were to reduce the clone time, and only download the binary file when needed which you current experienced. I believe without rewrite history, you end up with that same clone time, and still download full history of all files previously commited. Depend on your use case. If rewire needed, I would pause development, migrate, ask every one rebase their own branches to new "master". Am I correct @bk2204 ?

**Comment by bk2204** [comment]
If you want to rewrite history, then `git lfs migrate import` will do that.  If you don't want to rewrite history, like @phobos2077 has suggested, then it's fine to use either of the techniques mentioned, or to use `git lfs migrate import --no-rewrite` if that better meets your needs.

**Comment by phobos2077** [comment]
@nhatkhai you are correct, however I'm more concerned about future growth of the repository size. Currently it's around 5GB but the project is expected to enter a stage where a lot of new binary assets will be added and we want to avoid possible issues due to huge repository sizes. We don't need super fast clone time because it's not open source project and people clone it very occasionally, but we don't want to end up with cloning taking forever or git being too slow etc. Also git has a file size limit of 100MB which we already had an issue with.
